// toolbox/dist/nextjs/index.js
import {
  __spreadProps,
  __spreadValues
} from "../chunk-N3ETBM74.js";

// src/_frontend/hooks.ts
import { useState, useEffect } from "react";
var toolboxGet = async (url, action) => {
  const fetchUrl = new URL(url, window.location.origin);
  fetchUrl.searchParams.set("action", action);
  console.log("Fetching", fetchUrl);
  const res = await fetch(fetchUrl);
  return await res.json();
};
var toolboxPost = async (url, action, data) => {
  const fetchUrl = new URL(url, window.location.origin);
  fetchUrl.searchParams.set("action", action);
  const res = await fetch(fetchUrl, {
    method: "POST",
    body: data
  });
  return await res.json();
};
var useToolbox = (toolboxUrl) => {
  const url = toolboxUrl || "/toolbox";
  const [isLoaded, setIsLoaded] = useState(false);
  const [servers, setServers] = useState([]);
  const retryDiscovery = async (serverId) => {
    const formData = new FormData();
    formData.append("serverId", serverId);
    const res = await toolboxPost(url, "retry_discovery", formData);
    setServers(addMethodsToServers(res.servers));
  };
  const generateAuthorizationUrl = async (serverId, scopes, finalRedirectUrl) => {
    const formData = new FormData();
    formData.append("serverId", serverId);
    if (scopes) {
      formData.append("scopes", scopes.join(" "));
    }
    const redirectUrl = finalRedirectUrl || window.location.href;
    formData.append("finalRedirectUrl", redirectUrl);
    const res = await toolboxPost(url, "generate_authorization_url", formData);
    return res.authorizationUrl ? { authorizationUrl: res.authorizationUrl } : { error: res.error };
  };
  const addMethodsToServers = (servers2) => {
    return servers2.map((server) => __spreadProps(__spreadValues({}, server), {
      retryDiscovery: () => retryDiscovery(server.id),
      generateAuthorizationUrl: (scopes, finalRedirectUrl) => generateAuthorizationUrl(server.id, scopes, finalRedirectUrl)
    }));
  };
  useEffect(() => {
    async function fetchToolbox() {
      const res = await toolboxGet(url, "read_toolbox");
      setServers(addMethodsToServers(res.servers));
      setIsLoaded(true);
    }
    fetchToolbox();
  }, [url]);
  const toolbox = {
    servers,
    addServer: async (data) => {
      const res = await toolboxPost(url, "add_server", data);
      setServers(addMethodsToServers(res.servers));
    }
  };
  return { toolbox, isLoaded };
};
export {
  useToolbox
};
