// toolbox/dist/nextjs/server.js
import {
  __spreadProps,
  __spreadValues
} from "../chunk-N3ETBM74.js";

// src/_backend/toolbox-base.ts
import { randomUUID } from "crypto";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";

// src/_backend/streamableHttp.ts
import {
  isInitializedNotification,
  isJSONRPCRequest,
  isJSONRPCResponse,
  JSONRPCMessageSchema
} from "@modelcontextprotocol/sdk/types.js";
import {
  UnauthorizedError
} from "@modelcontextprotocol/sdk/client/auth.js";
import { EventSourceParserStream } from "eventsource-parser/stream";

// src/_backend/authorization-error.ts
var AuthorizationError = class extends Error {
  constructor(request, response) {
    const wwwAuthenticate = response.headers.get("WWW-Authenticate");
    super(
      `Request to ${request.url} returned 401. ${!wwwAuthenticate ? `The server did not provide a challenge from the WWW-Authenticate header.` : `The server provided the following challenge from the WWW-Authenticate header: ${wwwAuthenticate}`}`
    );
    this.request = request;
    this.response = response;
  }
};

// src/_backend/streamableHttp.ts
var DEFAULT_STREAMABLE_HTTP_RECONNECTION_OPTIONS = {
  initialReconnectionDelay: 1e3,
  maxReconnectionDelay: 3e4,
  reconnectionDelayGrowFactor: 1.5,
  maxRetries: 2
};
var StreamableHTTPError = class extends Error {
  constructor(code, message) {
    super(`Streamable HTTP error: ${message}`);
    this.code = code;
  }
};
var StreamableHTTPClientTransport = class {
  constructor(url, opts) {
    var _a;
    this._url = url;
    this._resourceMetadataUrl = void 0;
    this._requestInit = opts == null ? void 0 : opts.requestInit;
    this._authProvider = opts == null ? void 0 : opts.authProvider;
    this._sessionId = opts == null ? void 0 : opts.sessionId;
    this._reconnectionOptions = (_a = opts == null ? void 0 : opts.reconnectionOptions) != null ? _a : DEFAULT_STREAMABLE_HTTP_RECONNECTION_OPTIONS;
  }
  async _authThenStart() {
    var _a;
    if (!this._authProvider) {
      throw new UnauthorizedError("No auth provider");
    }
    let result;
    try {
      result = await auth(this._authProvider, {
        serverUrl: this._url,
        resourceMetadataUrl: this._resourceMetadataUrl
      });
    } catch (error) {
      (_a = this.onerror) == null ? void 0 : _a.call(this, error);
      throw error;
    }
    if (result !== "AUTHORIZED") {
      throw new UnauthorizedError();
    }
    return await this._startOrAuthSse({ resumptionToken: void 0 });
  }
  async _commonHeaders() {
    var _a;
    const headers = {};
    if (this._authProvider) {
      const accessToken = await this._authProvider.getAccessToken();
      if (accessToken) {
        headers["Authorization"] = `Bearer ${accessToken}`;
      }
    }
    if (this._sessionId) {
      headers["mcp-session-id"] = this._sessionId;
    }
    return new Headers(__spreadValues(__spreadValues({}, headers), (_a = this._requestInit) == null ? void 0 : _a.headers));
  }
  async _startOrAuthSse(options) {
    var _a, _b;
    const { resumptionToken } = options;
    try {
      const headers = await this._commonHeaders();
      headers.set("Accept", "text/event-stream");
      if (resumptionToken) {
        headers.set("last-event-id", resumptionToken);
      }
      const response = await fetch(this._url, {
        method: "GET",
        headers,
        signal: (_a = this._abortController) == null ? void 0 : _a.signal
      });
      if (!response.ok) {
        if (response.status === 401 && this._authProvider) {
          return await this._authThenStart();
        }
        if (response.status === 405) {
          return;
        }
        throw new StreamableHTTPError(
          response.status,
          `Failed to open SSE stream: ${response.statusText}`
        );
      }
      this._handleSseStream(response.body, options);
    } catch (error) {
      (_b = this.onerror) == null ? void 0 : _b.call(this, error);
      throw error;
    }
  }
  /**
   * Calculates the next reconnection delay using  backoff algorithm
   *
   * @param attempt Current reconnection attempt count for the specific stream
   * @returns Time to wait in milliseconds before next reconnection attempt
   */
  _getNextReconnectionDelay(attempt) {
    const initialDelay = this._reconnectionOptions.initialReconnectionDelay;
    const growFactor = this._reconnectionOptions.reconnectionDelayGrowFactor;
    const maxDelay = this._reconnectionOptions.maxReconnectionDelay;
    return Math.min(initialDelay * Math.pow(growFactor, attempt), maxDelay);
  }
  /**
   * Schedule a reconnection attempt with exponential backoff
   *
   * @param lastEventId The ID of the last received event for resumability
   * @param attemptCount Current reconnection attempt count for this specific stream
   */
  _scheduleReconnection(options, attemptCount = 0) {
    var _a;
    const maxRetries = this._reconnectionOptions.maxRetries;
    if (maxRetries > 0 && attemptCount >= maxRetries) {
      (_a = this.onerror) == null ? void 0 : _a.call(
        this,
        new Error(`Maximum reconnection attempts (${maxRetries}) exceeded.`)
      );
      return;
    }
    const delay = this._getNextReconnectionDelay(attemptCount);
    setTimeout(() => {
      this._startOrAuthSse(options).catch((error) => {
        var _a2;
        (_a2 = this.onerror) == null ? void 0 : _a2.call(
          this,
          new Error(
            `Failed to reconnect SSE stream: ${error instanceof Error ? error.message : String(error)}`
          )
        );
        this._scheduleReconnection(options, attemptCount + 1);
      });
    }, delay);
  }
  _handleSseStream(stream, options) {
    if (!stream) {
      return;
    }
    const { onresumptiontoken, replayMessageId } = options;
    let lastEventId;
    const processStream = async () => {
      var _a, _b, _c, _d;
      try {
        const reader = stream.pipeThrough(new TextDecoderStream()).pipeThrough(new EventSourceParserStream()).getReader();
        while (true) {
          const { value: event, done } = await reader.read();
          if (done) {
            break;
          }
          if (event.id) {
            lastEventId = event.id;
            onresumptiontoken == null ? void 0 : onresumptiontoken(event.id);
          }
          if (!event.event || event.event === "message") {
            try {
              const message = JSONRPCMessageSchema.parse(
                JSON.parse(event.data)
              );
              if (replayMessageId !== void 0 && isJSONRPCResponse(message)) {
                message.id = replayMessageId;
              }
              (_a = this.onmessage) == null ? void 0 : _a.call(this, message);
            } catch (error) {
              (_b = this.onerror) == null ? void 0 : _b.call(this, error);
            }
          }
        }
      } catch (error) {
        (_c = this.onerror) == null ? void 0 : _c.call(this, new Error(`SSE stream disconnected: ${error}`));
        if (this._abortController && !this._abortController.signal.aborted) {
          if (lastEventId !== void 0) {
            try {
              this._scheduleReconnection(
                {
                  resumptionToken: lastEventId,
                  onresumptiontoken,
                  replayMessageId
                },
                0
              );
            } catch (error2) {
              (_d = this.onerror) == null ? void 0 : _d.call(
                this,
                new Error(
                  `Failed to reconnect: ${error2 instanceof Error ? error2.message : String(error2)}`
                )
              );
            }
          }
        }
      }
    };
    processStream();
  }
  async start() {
    if (this._abortController) {
      throw new Error(
        "StreamableHTTPClientTransport already started! If using Client class, note that connect() calls start() automatically."
      );
    }
    this._abortController = new AbortController();
  }
  /**
   * Call this method after the user has finished authorizing via their user agent and is redirected back to the MCP client application. This will exchange the authorization code for an access token, enabling the next connection attempt to successfully auth.
   */
  async finishAuth(authorizationCode) {
    if (!this._authProvider) {
      throw new UnauthorizedError("No auth provider");
    }
    const result = await auth(this._authProvider, {
      serverUrl: this._url,
      authorizationCode,
      resourceMetadataUrl: this._resourceMetadataUrl
    });
    if (result !== "AUTHORIZED") {
      throw new UnauthorizedError("Failed to authorize");
    }
  }
  async close() {
    var _a, _b;
    (_a = this._abortController) == null ? void 0 : _a.abort();
    (_b = this.onclose) == null ? void 0 : _b.call(this);
  }
  async send(message, options) {
    var _a, _b, _c;
    try {
      const { resumptionToken, onresumptiontoken } = options || {};
      if (resumptionToken) {
        this._startOrAuthSse({
          resumptionToken,
          replayMessageId: isJSONRPCRequest(message) ? message.id : void 0
        }).catch((err) => {
          var _a2;
          return (_a2 = this.onerror) == null ? void 0 : _a2.call(this, err);
        });
        return;
      }
      const headers = await this._commonHeaders();
      headers.set("content-type", "application/json");
      headers.set("accept", "application/json, text/event-stream");
      const init = __spreadProps(__spreadValues({}, this._requestInit), {
        method: "POST",
        headers,
        body: JSON.stringify(message),
        signal: (_a = this._abortController) == null ? void 0 : _a.signal
      });
      const request = new Request(this._url, init);
      const response = await fetch(request);
      const sessionId = response.headers.get("mcp-session-id");
      if (sessionId) {
        this._sessionId = sessionId;
      }
      if (!response.ok) {
        if (response.status === 401) {
          if (this._authProvider) {
            await this._authProvider.refreshAccessToken();
            return this.send(message);
          } else {
            throw new AuthorizationError(request, response);
          }
        }
        const text = await response.text().catch(() => null);
        throw new Error(
          `Error POSTing to endpoint (HTTP ${response.status}): ${text}`
        );
      }
      if (response.status === 202) {
        if (isInitializedNotification(message)) {
          this._startOrAuthSse({ resumptionToken: void 0 }).catch(
            (err) => {
              var _a2;
              return (_a2 = this.onerror) == null ? void 0 : _a2.call(this, err);
            }
          );
        }
        return;
      }
      const messages = Array.isArray(message) ? message : [message];
      const hasRequests = messages.filter(
        (msg) => "method" in msg && "id" in msg && msg.id !== void 0
      ).length > 0;
      const contentType = response.headers.get("content-type");
      if (hasRequests) {
        if (contentType == null ? void 0 : contentType.includes("text/event-stream")) {
          this._handleSseStream(response.body, { onresumptiontoken });
        } else if (contentType == null ? void 0 : contentType.includes("application/json")) {
          const data = await response.json();
          const responseMessages = Array.isArray(data) ? data.map((msg) => JSONRPCMessageSchema.parse(msg)) : [JSONRPCMessageSchema.parse(data)];
          for (const msg of responseMessages) {
            (_b = this.onmessage) == null ? void 0 : _b.call(this, msg);
          }
        } else {
          throw new StreamableHTTPError(
            -1,
            `Unexpected content type: ${contentType}`
          );
        }
      }
    } catch (error) {
      (_c = this.onerror) == null ? void 0 : _c.call(this, error);
      throw error;
    }
  }
  get sessionId() {
    return this._sessionId;
  }
  /**
   * Terminates the current session by sending a DELETE request to the server.
   *
   * Clients that no longer need a particular session
   * (e.g., because the user is leaving the client application) SHOULD send an
   * HTTP DELETE to the MCP endpoint with the Mcp-Session-Id header to explicitly
   * terminate the session.
   *
   * The server MAY respond with HTTP 405 Method Not Allowed, indicating that
   * the server does not allow clients to terminate sessions.
   */
  async terminateSession() {
    var _a, _b;
    if (!this._sessionId) {
      return;
    }
    try {
      const headers = await this._commonHeaders();
      const init = __spreadProps(__spreadValues({}, this._requestInit), {
        method: "DELETE",
        headers,
        signal: (_a = this._abortController) == null ? void 0 : _a.signal
      });
      const response = await fetch(this._url, init);
      if (!response.ok && response.status !== 405) {
        throw new StreamableHTTPError(
          response.status,
          `Failed to terminate session: ${response.statusText}`
        );
      }
      this._sessionId = void 0;
    } catch (error) {
      (_b = this.onerror) == null ? void 0 : _b.call(this, error);
      throw error;
    }
  }
};

// src/_backend/toolbox-base.ts
var ToolboxBase = class {
  constructor({ contextId, store, auth: auth2 }) {
    this.transports = {};
    this.contextId = contextId != null ? contextId : auth2.userId;
    this.store = store;
    this.auth = auth2;
  }
  async successResponse() {
    return await Response.json(await this.readFullToolboxContext());
  }
  async handleGet(request) {
    const url = new URL(request.url);
    const searchParams = url.searchParams;
    if (searchParams.get("action") === "read_toolbox") {
      return await this.successResponse();
    }
    if (searchParams.get("action") === "oauth_callback") {
      if (this.auth.handleAuthorizationCallback) {
        const toolboxCallbackUrl = this.computeToolboxCallbackUrl(request);
        const result = await this.auth.handleAuthorizationCallback(
          request,
          toolboxCallbackUrl
        );
        return Response.redirect(result.finalRedirectUrl);
      } else {
        return Response.json({ error: "Not implemented" }, { status: 400 });
      }
    }
    return Response.json({ error: "Unknown action" }, { status: 400 });
  }
  async handlePost(request) {
    const url = new URL(request.url);
    const searchParams = url.searchParams;
    const requestBody = await request.formData();
    if (searchParams.get("action") === "add_server") {
      const toolboxCallbackUrl = this.computeToolboxCallbackUrl(request);
      await this.addMcpServer({
        endpoint: requestBody.get("url"),
        toolboxCallbackUrl
      });
      return await this.successResponse();
    }
    if (searchParams.get("action") === "retry_discovery") {
      const serverId = requestBody.get("serverId");
      if (!serverId) {
        return Response.json(
          { error: "serverId is required" },
          { status: 400 }
        );
      }
      const toolboxCallbackUrl = this.computeToolboxCallbackUrl(request);
      await this.retryDiscovery(serverId, toolboxCallbackUrl);
      return await this.successResponse();
    }
    if (searchParams.get("action") === "generate_authorization_url") {
      const serverId = requestBody.get("serverId");
      const scopes = requestBody.get("scopes");
      const finalRedirectUrl = requestBody.get("finalRedirectUrl");
      if (!serverId) {
        return Response.json(
          { error: "serverId is required" },
          { status: 400 }
        );
      }
      if (!finalRedirectUrl) {
        return Response.json(
          { error: "finalRedirectUrl is required" },
          { status: 400 }
        );
      }
      const toolboxCallbackUrl = this.computeToolboxCallbackUrl(request);
      const isValidFinalRedirectUrl = await this.auth.validateFinalRedirectUrl(
        finalRedirectUrl,
        toolboxCallbackUrl
      );
      if (!isValidFinalRedirectUrl) {
        return Response.json(
          {
            error: `Final redirect URL ${finalRedirectUrl} is not valid for toolbox callback URL ${toolboxCallbackUrl}. URLs must be same-site.`
          },
          { status: 400 }
        );
      }
      const result = await this.generateAuthorizationUrl({
        serverId,
        scopes: scopes ? scopes.split(" ") : void 0,
        toolboxCallbackUrl,
        finalRedirectUrl
      });
      if ("error" in result) {
        return Response.json({ error: result.error }, { status: 400 });
      }
      return Response.json({ authorizationUrl: result.authorizationUrl });
    }
    return Response.json({ error: "Unknown action" }, { status: 400 });
  }
  getScopedStorageKey(key) {
    return `toolbox_${this.contextId}_${key}`;
  }
  computeToolboxCallbackUrl(request) {
    const toolboxCallbackUrl = new URL(request.url);
    console.log("Starting with", toolboxCallbackUrl.toString());
    toolboxCallbackUrl.search = "";
    toolboxCallbackUrl.searchParams.set("action", "oauth_callback");
    console.log("Returning", toolboxCallbackUrl.toString());
    return toolboxCallbackUrl.toString();
  }
  async readFullToolboxContext() {
    const serverList = await this.readServerList();
    const servers = await Promise.all(
      serverList.map(async (server) => {
        const oauthMetadata = server.authorization.authorizationServerMetadata;
        const computedStatus = await this.computeAuthorizationStatus(server);
        return {
          id: server.id,
          url: server.url,
          authorization: __spreadProps(__spreadValues({}, server.authorization), {
            status: computedStatus
          }),
          hasAccessToken: oauthMetadata ? await this.auth.hasAccessToken({
            authorizationServerUrl: oauthMetadata.authorization_endpoint
          }) : false
        };
      })
    );
    return {
      servers
    };
  }
  async computeAuthorizationStatus(server) {
    if (server.authorization.unauthedStatusCode === 200) {
      return "not_required";
    }
    if (server.authorization.authorizationServerError !== null || server.authorization.protectedResourceMetadataError !== null) {
      return "discovery_failed";
    }
    if (!server.authorization.authorizationServerMetadata) {
      return "discovery_failed";
    }
    const oauthMetadata = server.authorization.authorizationServerMetadata;
    const hasAccessToken = await this.auth.hasAccessToken({
      authorizationServerUrl: oauthMetadata.authorization_endpoint
    });
    if (hasAccessToken) {
      return "connected";
    }
    const isRegistered = await this.auth.isRegistered({
      authorizationServerUrl: oauthMetadata.authorization_endpoint
    });
    if (isRegistered) {
      return "requires_connection";
    }
    return "requires_registration";
  }
  async readServerList() {
    return await this.store.read(
      this.getScopedStorageKey("serverlist")
    ) || [];
  }
  async writeServerList(servers) {
    return await this.store.write(
      this.getScopedStorageKey("serverlist"),
      servers
    );
  }
  async readServer(serverId) {
    return await this.store.read(
      this.getScopedStorageKey(`server_${serverId}`)
    );
  }
  async writeServer(serverId, server) {
    return await this.store.write(
      this.getScopedStorageKey(`server_${serverId}`),
      server
    );
  }
  async addMcpServer({
    endpoint,
    toolboxCallbackUrl
  }) {
    let serverList = await this.readServerList();
    if (serverList.some((server) => server.url === endpoint)) {
      return;
    }
    const serverId = randomUUID();
    const serverData = {
      id: serverId,
      url: endpoint,
      authorization: {
        protectedResourceMetadata: {},
        protectedResourceMetadataError: null,
        authorizationServerUrl: "",
        authorizationServerMetadata: {},
        authorizationServerError: null,
        unauthedWwwAuthenticate: null,
        unauthedStatusCode: null,
        registrationError: null
      }
    };
    await this.attemptConnectionAndDiscovery(serverData);
    if (serverData.authorization.authorizationServerMetadata && !serverData.authorization.authorizationServerError && !serverData.authorization.protectedResourceMetadataError && toolboxCallbackUrl) {
      const status = await this.computeAuthorizationStatus(serverData);
      if (status === "requires_registration") {
        const result = await this.registerClient(
          serverData,
          toolboxCallbackUrl
        );
        if ("error" in result) {
          serverData.authorization.registrationError = result.error;
        }
      }
    }
    await this.writeServerList([...serverList, serverData]);
    await this.writeServer(serverId, serverData);
    return;
  }
  async generateAuthorizationUrl({
    serverId,
    scopes,
    toolboxCallbackUrl,
    finalRedirectUrl
  }) {
    const serverData = await this.readServer(serverId);
    if (!serverData || !serverData.authorization.authorizationServerMetadata) {
      return { error: "Could not find authorization server" };
    }
    return this.auth.generateAuthorizationUrl({
      oauthMetadata: serverData.authorization.authorizationServerMetadata,
      scopes,
      toolboxCallbackUrl,
      finalRedirectUrl
    });
  }
  async registerClient(serverData, toolboxCallbackUrl) {
    if (!serverData.authorization.authorizationServerMetadata) {
      return { error: "Could not find authorization server" };
    }
    const result = await this.auth.registerDynamically({
      oauthMetadata: serverData.authorization.authorizationServerMetadata,
      toolboxCallbackUrl
    });
    if ("error" in result) {
      return { error: result.error };
    }
    await this.writeServer(serverData.id, serverData);
    const serverList = await this.readServerList();
    const updatedServerList = serverList.map(
      (server) => server.id === serverData.id ? serverData : server
    );
    await this.writeServerList(updatedServerList);
    return { success: true };
  }
  async getTransport(mcpServerUrl) {
    if (this.transports[mcpServerUrl]) {
      return this.transports[mcpServerUrl];
    }
    const serverData = await this.findServerByUrl(mcpServerUrl);
    if (!serverData || !serverData.authorization.authorizationServerMetadata) {
      throw new Error("Could not find authorization server");
    }
    const oauthMetadata = serverData.authorization.authorizationServerMetadata;
    this.transports[mcpServerUrl] = new StreamableHTTPClientTransport(
      new URL(mcpServerUrl),
      {
        authProvider: {
          getAccessToken: () => {
            return this.auth.getAccessToken({
              authorizationServerUrl: oauthMetadata.authorization_endpoint
            });
          },
          refreshAccessToken: async () => {
            await this.auth.getAccessToken({
              authorizationServerUrl: oauthMetadata.authorization_endpoint
            });
          }
        }
      }
    );
    return this.transports[mcpServerUrl];
  }
  async callTool({
    mcpServerUrl,
    toolName,
    toolArguments
  }) {
    const serverData = await this.findServerByUrl(mcpServerUrl);
    if (!serverData || !serverData.authorization.authorizationServerMetadata) {
      throw new Error("Could not find authorization server");
    }
    const mcpClient = new Client({
      name: "client.mcpClientName",
      version: "0.0.1"
    });
    const transport = await this.getTransport(mcpServerUrl);
    await mcpClient.connect(transport);
    return await mcpClient.callTool({
      name: toolName,
      arguments: toolArguments
    });
  }
  async removeMcpServer({ endpoint }) {
  }
  async listMcpServers() {
  }
  async listTools() {
  }
  async findServerByUrl(url) {
    const serverList = await this.readServerList();
    return serverList.find((server) => server.url === url);
  }
  async retryDiscovery(serverId, toolboxCallbackUrl) {
    const serverData = await this.readServer(serverId);
    if (!serverData) {
      throw new Error(`Server with ID ${serverId} not found`);
    }
    console.log("Retrying discovery for", serverData.url);
    serverData.authorization.protectedResourceMetadata = {};
    serverData.authorization.protectedResourceMetadataError = null;
    serverData.authorization.authorizationServerUrl = "";
    serverData.authorization.authorizationServerMetadata = {};
    serverData.authorization.authorizationServerError = null;
    serverData.authorization.unauthedWwwAuthenticate = null;
    serverData.authorization.unauthedStatusCode = null;
    serverData.authorization.registrationError = null;
    await this.attemptConnectionAndDiscovery(serverData);
    if (serverData.authorization.authorizationServerMetadata && !serverData.authorization.authorizationServerError && !serverData.authorization.protectedResourceMetadataError) {
      const status = await this.computeAuthorizationStatus(serverData);
      if (status === "requires_registration") {
        const result = await this.registerClient(
          serverData,
          toolboxCallbackUrl
        );
        if ("error" in result) {
          serverData.authorization.registrationError = result.error;
        }
      }
    }
    console.log("End retry discovery attempt", serverData.url);
    const serverList = await this.readServerList();
    const updatedServerList = serverList.map(
      (server) => server.id === serverId ? serverData : server
    );
    await this.writeServerList(updatedServerList);
    await this.writeServer(serverId, serverData);
    return;
  }
  async attemptConnectionAndDiscovery(serverData) {
    console.log("Connecting to", serverData.url);
    const mcpClient = new Client({
      name: "client.mcpClientName",
      version: "0.0.1"
    });
    const transport = new StreamableHTTPClientTransport(
      new URL(serverData.url)
    );
    try {
      await mcpClient.connect(transport);
      serverData.authorization.protectedResourceMetadata = {};
      serverData.authorization.protectedResourceMetadataError = null;
      serverData.authorization.authorizationServerUrl = "";
      serverData.authorization.authorizationServerMetadata = {};
      serverData.authorization.authorizationServerError = null;
      serverData.authorization.unauthedWwwAuthenticate = null;
      serverData.authorization.unauthedStatusCode = 200;
      console.log("Connection successful - no auth required");
    } catch (e) {
      if (e instanceof AuthorizationError) {
        console.log(
          "Authorization error - auth required",
          e.message,
          e.response,
          e.request
        );
        const wwwAuthenticate = e.response.headers.get("WWW-Authenticate");
        serverData.authorization.unauthedWwwAuthenticate = wwwAuthenticate;
        serverData.authorization.unauthedStatusCode = e.response.status || null;
        if (wwwAuthenticate) {
          console.log(
            "Attempting to discover authorization server from WWW-Authenticate header"
          );
          try {
            const discoveryResult = await this.auth.discoverAuthorizationServerFromWWWAuthenticate(
              wwwAuthenticate,
              serverData.url
            );
            serverData.authorization.authorizationServerUrl = discoveryResult.authorizationServerUrl;
            serverData.authorization.protectedResourceMetadata = discoveryResult.protectedResourceMetadata;
            serverData.authorization.protectedResourceMetadataError = discoveryResult.protectedResourceMetadataError;
            serverData.authorization.authorizationServerMetadata = discoveryResult.authorizationServerMetadata;
            serverData.authorization.authorizationServerError = discoveryResult.authorizationServerError;
            if (!discoveryResult.protectedResourceMetadataError && !discoveryResult.authorizationServerError) {
              console.log(
                "Successfully discovered authorization server:",
                discoveryResult.authorizationServerUrl
              );
            } else {
              console.warn(
                "Discovery completed with errors:",
                discoveryResult.protectedResourceMetadataError || discoveryResult.authorizationServerError
              );
            }
          } catch (discoveryError) {
            console.warn("Error during metadata discovery:", discoveryError);
            serverData.authorization.authorizationServerError = `Discovery error: ${discoveryError}`;
          }
        } else {
          console.log(
            "No WWW-Authenticate header found, cannot discover metadata"
          );
          serverData.authorization.authorizationServerError = "No WWW-Authenticate header provided";
        }
      } else {
        console.log("Connection failed - auth may be required", e);
        serverData.authorization.authorizationServerError = "Connection failed - not an authorization error";
        serverData.authorization.unauthedWwwAuthenticate = null;
        serverData.authorization.unauthedStatusCode = null;
      }
    }
  }
};

// src/nextjs/toolbox.ts
var Toolbox = class extends ToolboxBase {
  constructor(props) {
    super(__spreadValues({}, props));
  }
  handlers() {
    return {
      GET: this.handleGet.bind(this),
      POST: this.handlePost.bind(this)
    };
  }
};
export {
  Toolbox
};
