// toolbox/dist/nextjs/auth/self-storage.js
import {
  __spreadValues
} from "../../chunk-N3ETBM74.js";

// src/nextjs/auth/self-storage.ts
import {
  discoverOAuthMetadata,
  exchangeAuthorization,
  registerClient,
  startAuthorization
} from "@modelcontextprotocol/sdk/client/auth.js";
import { randomUUID } from "crypto";

// src/_backend/auth/provider-base.ts
import * as psl from "psl";
var AuthProviderBase = class {
  constructor({
    discoveryOptions
  }) {
    this.discoveryOptions = __spreadValues({
      resourceVsProtectedResourceMetadata: "strict-spec",
      resourceVsAuthorizationServer: "same-site"
    }, discoveryOptions);
  }
  async validateResourceVsProtectedResourceMetadata(resourceUri, protectedResourceMetadataUri) {
    console.log(
      "Validating resource vs protected resource metadata",
      this.discoveryOptions.resourceVsProtectedResourceMetadata,
      resourceUri,
      protectedResourceMetadataUri
    );
    if (this.discoveryOptions.resourceVsProtectedResourceMetadata === "any") {
      return true;
    }
    if (this.discoveryOptions.resourceVsProtectedResourceMetadata === "same-site") {
      const resourceUrl = new URL(resourceUri);
      const metadataUrl = new URL(protectedResourceMetadataUri);
      return psl.get(resourceUrl.hostname) === psl.get(metadataUrl.hostname);
    }
    if (this.discoveryOptions.resourceVsProtectedResourceMetadata === "same-origin") {
      const resourceUrl = new URL(resourceUri);
      const metadataUrl = new URL(protectedResourceMetadataUri);
      return resourceUrl.origin === metadataUrl.origin;
    }
    if (this.discoveryOptions.resourceVsProtectedResourceMetadata === "strict-spec") {
      const metadataUrl = new URL(protectedResourceMetadataUri);
      const expectedMetadataUrl = new URL(resourceUri);
      const pathWithoutTrailingSlash = expectedMetadataUrl.pathname.replace(
        /\/$/,
        ""
      );
      if (pathWithoutTrailingSlash === "") {
        expectedMetadataUrl.pathname = "/.well-known/oauth-protected-resource";
      } else {
        expectedMetadataUrl.pathname = `/.well-known/oauth-protected-resource${pathWithoutTrailingSlash}`;
      }
      return metadataUrl.href === expectedMetadataUrl.href;
    }
    if (typeof this.discoveryOptions.resourceVsProtectedResourceMetadata === "function") {
      return this.discoveryOptions.resourceVsProtectedResourceMetadata(
        resourceUri,
        protectedResourceMetadataUri
      );
    }
  }
  async validateResourceVsAuthorizationServer(resourceUri, authorizationServerUri) {
    if (this.discoveryOptions.resourceVsAuthorizationServer === "any") {
      return true;
    }
    if (this.discoveryOptions.resourceVsAuthorizationServer === "same-site") {
      const resourceUrl = new URL(resourceUri);
      const authServerUrl = new URL(authorizationServerUri);
      return psl.get(resourceUrl.hostname) === psl.get(authServerUrl.hostname);
    }
    if (this.discoveryOptions.resourceVsAuthorizationServer === "same-origin") {
      const resourceUrl = new URL(resourceUri);
      const authServerUrl = new URL(authorizationServerUri);
      return resourceUrl.origin === authServerUrl.origin;
    }
    if (typeof this.discoveryOptions.resourceVsAuthorizationServer === "function") {
      return this.discoveryOptions.resourceVsAuthorizationServer(
        resourceUri,
        authorizationServerUri
      );
    }
  }
  async validateFinalRedirectUrl(finalRedirectUrl, toolboxCallbackUrl) {
    try {
      const finalUrl = new URL(finalRedirectUrl);
      const callbackUrl = new URL(toolboxCallbackUrl);
      return psl.get(finalUrl.hostname) === psl.get(callbackUrl.hostname);
    } catch (error) {
      return false;
    }
  }
  /**
   * Discovers the authorization server URL from a WWW-Authenticate header
   * based on RFC 9728, but uses the existing validation methods instead of
   * the precise spec.
   *
   * @param wwwAuthenticateHeader - The WWW-Authenticate header value
   * @param resourceUri - The URI of the protected resource that returned the header
   * @returns Object containing authorization server URL, protected resource metadata, and authorization server metadata, with error fields populated if discovery fails
   */
  async discoverAuthorizationServerFromWWWAuthenticate(wwwAuthenticateHeader, resourceUri) {
    console.log(
      "Discovering authorization server from the WWW-Authenticate header"
    );
    const authScheme = wwwAuthenticateHeader.split(" ")[0];
    if (authScheme.toLowerCase() !== "bearer") {
      return {
        authorizationServerUrl: "",
        protectedResourceMetadata: {},
        protectedResourceMetadataError: "WWW-Authenticate header does not use Bearer scheme",
        authorizationServerMetadata: {},
        authorizationServerError: null
      };
    }
    const params = this.parseWWWAuthenticateParams(wwwAuthenticateHeader);
    const resourceMetadataUrl = params.get("resource_metadata");
    if (!resourceMetadataUrl) {
      return {
        authorizationServerUrl: "",
        protectedResourceMetadata: {},
        protectedResourceMetadataError: "WWW-Authenticate header does not contain resource_metadata parameter",
        authorizationServerMetadata: {},
        authorizationServerError: null
      };
    }
    const isValidMetadata = await this.validateResourceVsProtectedResourceMetadata(
      resourceUri,
      resourceMetadataUrl
    );
    if (!isValidMetadata) {
      const currentOption = this.discoveryOptions.resourceVsProtectedResourceMetadata;
      return {
        authorizationServerUrl: "",
        protectedResourceMetadata: {},
        protectedResourceMetadataError: `Protected resource metadata URL ${resourceMetadataUrl} is not valid for resource ${resourceUri} with current discovery option "${currentOption}". Consider changing the 'resourceVsProtectedResourceMetadata' discovery option, though this may not be safe as it bypasses validation.`,
        authorizationServerMetadata: {},
        authorizationServerError: null
      };
    }
    try {
      const response = await fetch(resourceMetadataUrl);
      if (!response.ok) {
        return {
          authorizationServerUrl: "",
          protectedResourceMetadata: {},
          protectedResourceMetadataError: `Failed to fetch protected resource metadata from ${resourceMetadataUrl}: ${response.status} ${response.statusText}`,
          authorizationServerMetadata: {},
          authorizationServerError: null
        };
      }
      const protectedResourceMetadata = await response.json();
      const authorizationServers = protectedResourceMetadata.authorization_servers;
      if (!Array.isArray(authorizationServers) || authorizationServers.length === 0) {
        return {
          authorizationServerUrl: "",
          protectedResourceMetadata: {},
          protectedResourceMetadataError: "Protected resource metadata does not contain any authorization_servers",
          authorizationServerMetadata: {},
          authorizationServerError: null
        };
      }
      for (const authServerUrl of authorizationServers) {
        const isValidAuthServer = await this.validateResourceVsAuthorizationServer(
          resourceUri,
          authServerUrl
        );
        if (isValidAuthServer) {
          try {
            const authServerResponse = await fetch(
              `${authServerUrl}/.well-known/oauth-authorization-server`
            );
            if (!authServerResponse.ok) {
              console.warn(
                `Authorization server metadata not available at ${authServerUrl}/.well-known/oauth-authorization-server: ${authServerResponse.status} ${authServerResponse.statusText}`
              );
              continue;
            }
            const authorizationServerMetadata = await authServerResponse.json();
            return {
              authorizationServerUrl: authServerUrl,
              protectedResourceMetadata,
              protectedResourceMetadataError: null,
              authorizationServerMetadata,
              authorizationServerError: null
            };
          } catch (error) {
            console.warn(
              `Failed to fetch authorization server metadata from ${authServerUrl}:`,
              error
            );
            continue;
          }
        }
      }
      const currentOption = this.discoveryOptions.resourceVsAuthorizationServer;
      return {
        authorizationServerUrl: "",
        protectedResourceMetadata,
        protectedResourceMetadataError: null,
        authorizationServerMetadata: {},
        authorizationServerError: `No valid authorization server found. Tried ${authorizationServers.length} servers but none were valid or had accessible metadata with current discovery option "${currentOption}". Consider changing the 'resourceVsAuthorizationServer' discovery option, though this may not be safe as it bypasses validation.`
      };
    } catch (error) {
      return {
        authorizationServerUrl: "",
        protectedResourceMetadata: {},
        protectedResourceMetadataError: `Failed to discover authorization server from WWW-Authenticate header: ${error}`,
        authorizationServerMetadata: {},
        authorizationServerError: null
      };
    }
  }
  /**
   * Parses parameters from a WWW-Authenticate header value.
   * Handles quoted and unquoted parameter values.
   *
   * @param headerValue - The WWW-Authenticate header value
   * @returns Map of parameter names to values
   */
  parseWWWAuthenticateParams(headerValue) {
    const params = /* @__PURE__ */ new Map();
    const schemeEnd = headerValue.indexOf(" ");
    if (schemeEnd === -1) {
      return params;
    }
    const paramsString = headerValue.substring(schemeEnd + 1);
    const paramRegex = /(\w+)=([^,\s]+|"[^"]*")/g;
    let match;
    while ((match = paramRegex.exec(paramsString)) !== null) {
      const key = match[1];
      let value = match[2];
      if (value.startsWith('"') && value.endsWith('"')) {
        value = value.slice(1, -1);
      }
      params.set(key, value);
    }
    return params;
  }
};

// src/nextjs/auth/self-storage.ts
var SelfStorageAuthProvider = class extends AuthProviderBase {
  constructor({
    store,
    userId,
    oauthClientName,
    oauthClientUri,
    discoveryOptions
  }) {
    super({ discoveryOptions });
    this.store = store;
    this.oauthClientName = oauthClientName;
    this.oauthClientUri = oauthClientUri;
    this.scopes = ["openid", "profile", "email"];
    this.userId = userId;
  }
  getTokensStorageKey(server) {
    return `cm_tokens_${this.userId}_${server}`;
  }
  async getTokens(server) {
    const storageKey = this.getTokensStorageKey(server);
    const tokens = await this.store.read(storageKey);
    return tokens;
  }
  async writeTokens(server, tokens) {
    const storageKey = this.getTokensStorageKey(server);
    await this.store.write(storageKey, tokens);
  }
  getRegistrationStorageKey(server) {
    return `cm_registration_${server}`;
  }
  getStateStorageKey(state) {
    return `cm_state_${state}`;
  }
  async registerDynamically({
    oauthMetadata,
    toolboxCallbackUrl
  }) {
    const storageKey = this.getRegistrationStorageKey(
      oauthMetadata.authorization_endpoint
    );
    try {
      const dynamicRegistration = await registerClient(
        oauthMetadata.authorization_endpoint,
        {
          metadata: oauthMetadata,
          clientMetadata: {
            client_name: this.oauthClientName,
            client_uri: this.oauthClientUri,
            redirect_uris: [toolboxCallbackUrl]
          }
        }
      );
      await this.store.write(storageKey, dynamicRegistration);
      return dynamicRegistration;
    } catch (error) {
      console.error("Failed to register OAuth client:", error);
      return {
        error: error instanceof Error ? error.message : "Failed to register OAuth client"
      };
    }
  }
  // This is for SelfStorage, but if the OAuth manager is external like Clerk, you'd
  // just make an API request instead
  async getRegistrationData({
    oauthMetadata,
    toolboxCallbackUrl
  }) {
    const storageKey = this.getRegistrationStorageKey(
      oauthMetadata.authorization_endpoint
    );
    const registration = await this.store.read(storageKey);
    if (registration) {
      return registration;
    }
    try {
      const dyanamicRegistration = await registerClient(
        oauthMetadata.authorization_endpoint,
        {
          metadata: oauthMetadata,
          clientMetadata: {
            client_name: this.oauthClientName,
            client_uri: this.oauthClientUri,
            redirect_uris: [toolboxCallbackUrl]
          }
        }
      );
      await this.store.write(storageKey, dyanamicRegistration);
      return dyanamicRegistration;
    } catch (error) {
      console.error("Failed to register OAuth client:", error);
      return {
        error: error instanceof Error ? error.message : "Failed to register OAuth client"
      };
    }
  }
  async getStateData(state) {
    const stateStorageKey = this.getStateStorageKey(state);
    const stateData = await this.store.read(stateStorageKey);
    return stateData;
  }
  async generateAuthorizationUrl({
    oauthMetadata,
    scopes,
    toolboxCallbackUrl,
    finalRedirectUrl
  }) {
    const registrationData = await this.getRegistrationData({
      oauthMetadata,
      toolboxCallbackUrl
    });
    if ("error" in registrationData) {
      return {
        error: `Failed to get registration data: ${registrationData.error}`
      };
    }
    const state = randomUUID();
    const stateStorageKey = this.getStateStorageKey(state);
    try {
      const { authorizationUrl, codeVerifier } = await startAuthorization(
        oauthMetadata.authorization_endpoint,
        {
          metadata: oauthMetadata,
          clientInformation: registrationData,
          redirectUrl: toolboxCallbackUrl,
          scope: scopes == null ? void 0 : scopes.join(" "),
          state
        }
      );
      console.log("Generated with", toolboxCallbackUrl);
      await this.store.write(stateStorageKey, {
        authorizationEndpoint: oauthMetadata.authorization_endpoint,
        codeVerifier,
        finalRedirectUrl
      });
      return { authorizationUrl: authorizationUrl.href };
    } catch (error) {
      console.error("Failed to generate authorization URL:", error);
      return {
        error: error instanceof Error ? error.message : "Failed to generate authorization URL"
      };
    }
  }
  // Clerk won't need this
  async handleAuthorizationCallback(request, toolboxCallbackUrl) {
    const url = new URL(request.url);
    const state = url.searchParams.get("state");
    if (!state) {
      throw new Error("No state found");
    }
    const stateData = await this.getStateData(state);
    if (!stateData) {
      throw new Error("No state data found");
    }
    const authorizationCode = url.searchParams.get("code");
    if (!authorizationCode) {
      throw new Error("No authorization code found");
    }
    const registrationKey = await this.getRegistrationStorageKey(
      stateData.authorizationEndpoint
    );
    const registrationData = await this.store.read(
      registrationKey
    );
    const oauthMetadata = await discoverOAuthMetadata(
      stateData.authorizationEndpoint
    );
    console.log("OAUTH METADATA LOAD", oauthMetadata);
    console.log("REGISTRATION DATA LOAD", registrationData);
    const tokens = await exchangeAuthorization(
      stateData.authorizationEndpoint,
      {
        metadata: oauthMetadata,
        clientInformation: registrationData,
        authorizationCode,
        codeVerifier: stateData.codeVerifier,
        redirectUri: toolboxCallbackUrl
      }
    );
    await this.store.write(this.getStateStorageKey(state), null);
    await this.writeTokens(stateData.authorizationEndpoint, tokens);
    return { finalRedirectUrl: stateData.finalRedirectUrl };
  }
  async hasAccessToken({
    authorizationServerUrl
  }) {
    return !!await this.getAccessToken({ authorizationServerUrl });
  }
  async getAccessToken({
    authorizationServerUrl
  }) {
    const tokens = await this.getTokens(authorizationServerUrl);
    return tokens ? tokens.access_token : null;
  }
  async isRegistered({
    authorizationServerUrl
  }) {
    const storageKey = this.getRegistrationStorageKey(authorizationServerUrl);
    const registration = await this.store.read(storageKey);
    return !!registration;
  }
};
export {
  SelfStorageAuthProvider
};
