// toolbox/dist/storage/fs.js
import "../chunk-N3ETBM74.js";

// src/storage/fs.ts
import fs from "fs/promises";
import fsSync from "fs";
import os from "os";
import path from "path";
function createFsStore(config = {}) {
  const { filePath = path.join(os.tmpdir(), "__mcp_demo") } = config;
  const ensureDirectoryExists = async () => {
    const dir = path.dirname(filePath);
    try {
      await fs.mkdir(dir, { recursive: true });
    } catch (error) {
    }
  };
  const readStoreData = async () => {
    console.log("reading store data", filePath);
    try {
      if (!fsSync.existsSync(filePath)) {
        await ensureDirectoryExists();
        await fs.writeFile(filePath, "{}");
        return {};
      }
      const content = await fs.readFile(filePath, "utf8");
      return JSON.parse(content);
    } catch (error) {
      await ensureDirectoryExists();
      await fs.writeFile(filePath, "{}");
      return {};
    }
  };
  const writeStoreData = async (data) => {
    await ensureDirectoryExists();
    await fs.writeFile(filePath, JSON.stringify(data, null, 2));
  };
  const store = {
    write: async (k, v) => {
      var _a;
      const data = await readStoreData();
      data[k] = {
        value: v,
        created_at: ((_a = data[k]) == null ? void 0 : _a.created_at) || (/* @__PURE__ */ new Date()).toISOString(),
        updated_at: (/* @__PURE__ */ new Date()).toISOString()
      };
      await writeStoreData(data);
    },
    read: async (k) => {
      const data = await readStoreData();
      const entry = data[k];
      return entry ? entry.value : null;
    },
    /**
     * Delete a key from the store
     */
    delete: async (k) => {
      const data = await readStoreData();
      if (k in data) {
        delete data[k];
        await writeStoreData(data);
        return true;
      }
      return false;
    },
    /**
     * Check if a key exists in the store
     */
    exists: async (k) => {
      const data = await readStoreData();
      return k in data;
    },
    /**
     * List all keys in the store (useful for debugging)
     */
    keys: async () => {
      const data = await readStoreData();
      return Object.keys(data).sort((a, b) => {
        const aTime = new Date(data[a].created_at).getTime();
        const bTime = new Date(data[b].created_at).getTime();
        return aTime - bTime;
      });
    },
    /**
     * Clean up old entries (older than specified days)
     */
    cleanup: async (olderThanDays = 30) => {
      const data = await readStoreData();
      const cutoffDate = /* @__PURE__ */ new Date();
      cutoffDate.setDate(cutoffDate.getDate() - olderThanDays);
      let deleteCount = 0;
      for (const [key, entry] of Object.entries(data)) {
        const createdAt = new Date(entry.created_at);
        if (createdAt < cutoffDate) {
          delete data[key];
          deleteCount++;
        }
      }
      if (deleteCount > 0) {
        await writeStoreData(data);
      }
      return deleteCount;
    },
    /**
     * Get file system store statistics
     */
    stats: async () => {
      const data = await readStoreData();
      let fileSize = "0 B";
      try {
        if (fsSync.existsSync(filePath)) {
          const stats = await fs.stat(filePath);
          const bytes = stats.size;
          if (bytes < 1024) {
            fileSize = `${bytes} B`;
          } else if (bytes < 1024 * 1024) {
            fileSize = `${(bytes / 1024).toFixed(2)} KB`;
          } else {
            fileSize = `${(bytes / (1024 * 1024)).toFixed(2)} MB`;
          }
        }
      } catch (error) {
      }
      return {
        totalKeys: Object.keys(data).length,
        fileSize,
        filePath
      };
    },
    /**
     * No-op for file system store (no connection to close)
     */
    disconnect: async () => {
    }
  };
  return store;
}
var fsStore = createFsStore();
var fs_default = fsStore;
export {
  createFsStore,
  fs_default as default
};
