// toolbox/dist/storage/postgres.js
import "../chunk-N3ETBM74.js";

// src/storage/postgres.ts
import pg from "pg";
var { Client } = pg;
function createPostgresStore(config = {}) {
  if (!pg) {
    throw new Error(
      "pg package not found. Please install it with: npm install pg @types/pg"
    );
  }
  const {
    connectionString,
    host = "localhost",
    port = 5432,
    database,
    user,
    password,
    tableName = "mcp_store",
    ssl = false,
    connectionTimeoutMillis = 5e3,
    idleTimeoutMillis = 1e4,
    max = 10
  } = config;
  const clientConfig = connectionString ? { connectionString } : {
    host,
    port,
    database,
    user,
    password,
    ssl,
    connectionTimeoutMillis,
    idleTimeoutMillis,
    max
  };
  const client = new Client(clientConfig);
  client.on("error", (err) => {
    console.error("PostgreSQL Client Error:", err);
  });
  const initPromise = (async () => {
    await client.connect();
    await client.query(`
      CREATE TABLE IF NOT EXISTS ${tableName} (
        key VARCHAR(255) PRIMARY KEY,
        value JSONB NOT NULL,
        created_at TIMESTAMP WITH TIME ZONE DEFAULT NOW(),
        updated_at TIMESTAMP WITH TIME ZONE DEFAULT NOW()
      )
    `);
    await client.query(`
      CREATE INDEX IF NOT EXISTS idx_${tableName}_created_at 
      ON ${tableName} (created_at)
    `);
  })();
  const store = {
    write: async (k, v) => {
      await initPromise;
      const query = `
        INSERT INTO ${tableName} (key, value, updated_at) 
        VALUES ($1, $2, NOW()) 
        ON CONFLICT (key) 
        DO UPDATE SET value = $2, updated_at = NOW()
      `;
      await client.query(query, [k, JSON.stringify(v)]);
    },
    read: async (k) => {
      await initPromise;
      const query = `SELECT value FROM ${tableName} WHERE key = $1`;
      const result = await client.query(query, [k]);
      if (result.rows.length === 0) {
        return null;
      }
      const value = result.rows[0].value;
      if (typeof value === "object") {
        return value;
      }
      try {
        return JSON.parse(value);
      } catch (parseError) {
        throw new Error(
          `Failed to parse stored value for key "${k}": Invalid JSON`
        );
      }
    },
    /**
     * Delete a key from the store
     */
    delete: async (k) => {
      await initPromise;
      const query = `DELETE FROM ${tableName} WHERE key = $1`;
      const result = await client.query(query, [k]);
      return !!result.rowCount && result.rowCount > 0;
    },
    /**
     * Check if a key exists in the store
     */
    exists: async (k) => {
      await initPromise;
      const query = `SELECT 1 FROM ${tableName} WHERE key = $1`;
      const result = await client.query(query, [k]);
      return result.rows.length > 0;
    },
    /**
     * List all keys in the store (useful for debugging)
     */
    keys: async () => {
      await initPromise;
      const query = `SELECT key FROM ${tableName} ORDER BY created_at`;
      const result = await client.query(query);
      return result.rows.map((row) => row.key);
    },
    /**
     * Clean up old entries (older than specified days)
     */
    cleanup: async (olderThanDays = 30) => {
      var _a;
      await initPromise;
      const query = `
        DELETE FROM ${tableName} 
        WHERE created_at < NOW() - INTERVAL '${olderThanDays} days'
      `;
      const result = await client.query(query);
      return (_a = result.rowCount) != null ? _a : 0;
    },
    /**
     * Get PostgreSQL store statistics
     */
    stats: async () => {
      await initPromise;
      let totalKeys = 0;
      let tableSize = "N/A";
      let dbInfo = "N/A";
      try {
        const countResult = await client.query(
          `SELECT COUNT(*) as count FROM ${tableName}`
        );
        totalKeys = parseInt(countResult.rows[0].count);
        const sizeResult = await client.query(`
          SELECT pg_size_pretty(pg_total_relation_size('${tableName}')) as size
        `);
        if (sizeResult.rows.length > 0) {
          tableSize = sizeResult.rows[0].size;
        }
        const versionResult = await client.query("SELECT version()");
        if (versionResult.rows.length > 0) {
          const version = versionResult.rows[0].version;
          const match = version.match(/PostgreSQL ([\d.]+)/);
          if (match) {
            dbInfo = `PostgreSQL ${match[1]}`;
          }
        }
      } catch (error) {
      }
      return {
        totalKeys,
        tableSize,
        dbInfo
      };
    },
    /**
     * Close the PostgreSQL connection
     */
    disconnect: async () => {
      await client.end();
    }
  };
  return store;
}
export {
  createPostgresStore
};
