// toolbox/dist/storage/redis.js
import "../chunk-N3ETBM74.js";

// src/storage/redis.ts
import Redis from "redis";
function createRedisStore(config = {}) {
  if (!Redis) {
    throw new Error(
      "Redis package not found. Please install it with: npm install redis@^4.0.0"
    );
  }
  const {
    host = "localhost",
    port = 6379,
    password,
    db = 0,
    keyPrefix = "mcp:",
    connectTimeout = 5e3,
    ttl,
    tls = false
  } = config;
  const redisConfig = {
    socket: {
      host,
      port,
      connectTimeout,
      tls: tls ? {} : false
    },
    database: db
  };
  if (password) {
    redisConfig.password = password;
  }
  const redis = Redis.createClient(redisConfig);
  redis.on("error", (err) => {
    console.error("Redis Client Error:", err);
  });
  const formatKey = (key) => `${keyPrefix}${key}`;
  const connectionPromise = redis.connect();
  const store = {
    write: async (k, v) => {
      await connectionPromise;
      const serializedValue = JSON.stringify({
        value: v,
        created_at: (/* @__PURE__ */ new Date()).toISOString(),
        updated_at: (/* @__PURE__ */ new Date()).toISOString()
      });
      const key = formatKey(k);
      if (ttl) {
        await redis.setEx(key, ttl, serializedValue);
      } else {
        await redis.set(key, serializedValue);
      }
    },
    read: async (k) => {
      await connectionPromise;
      const key = formatKey(k);
      const value = await redis.get(key);
      if (value === null) {
        return null;
      }
      try {
        const parsed = JSON.parse(value);
        return parsed.value !== void 0 ? parsed.value : parsed;
      } catch (parseError) {
        throw new Error(
          `Failed to parse stored value for key "${k}": Invalid JSON`
        );
      }
    },
    /**
     * Delete a key from the store
     */
    delete: async (k) => {
      await connectionPromise;
      const key = formatKey(k);
      const result = await redis.del(key);
      return result > 0;
    },
    /**
     * Check if a key exists in the store
     */
    exists: async (k) => {
      await connectionPromise;
      const key = formatKey(k);
      const result = await redis.exists(key);
      return result === 1;
    },
    /**
     * List all keys in the store (useful for debugging)
     */
    keys: async () => {
      await connectionPromise;
      const pattern = `${keyPrefix}*`;
      const keys = await redis.keys(pattern);
      const unprefixedKeys = keys.map((key) => key.substring(keyPrefix.length));
      try {
        const keysWithTimes = await Promise.all(
          unprefixedKeys.map(async (key) => {
            try {
              const value = await redis.get(formatKey(key));
              if (value) {
                const parsed = JSON.parse(value);
                return {
                  key,
                  created_at: parsed.created_at || (/* @__PURE__ */ new Date(0)).toISOString()
                };
              }
            } catch (e) {
            }
            return { key, created_at: (/* @__PURE__ */ new Date(0)).toISOString() };
          })
        );
        return keysWithTimes.sort(
          (a, b) => new Date(a.created_at).getTime() - new Date(b.created_at).getTime()
        ).map((item) => item.key);
      } catch (e) {
        return unprefixedKeys;
      }
    },
    /**
     * Clean up old entries (older than specified days)
     * Note: This requires scanning all keys, which can be expensive for large datasets
     */
    cleanup: async (olderThanDays = 30) => {
      await connectionPromise;
      const pattern = `${keyPrefix}*`;
      const keys = await redis.keys(pattern);
      const cutoffDate = /* @__PURE__ */ new Date();
      cutoffDate.setDate(cutoffDate.getDate() - olderThanDays);
      let deleteCount = 0;
      for (const key of keys) {
        try {
          const value = await redis.get(key);
          if (value) {
            const parsed = JSON.parse(value);
            const createdAt = new Date(parsed.created_at || 0);
            if (createdAt < cutoffDate) {
              const deleted = await redis.del(key);
              if (deleted > 0) {
                deleteCount++;
              }
            }
          }
        } catch (e) {
        }
      }
      return deleteCount;
    },
    /**
     * Get Redis store statistics
     */
    stats: async () => {
      await connectionPromise;
      const pattern = `${keyPrefix}*`;
      const keys = await redis.keys(pattern);
      let memoryUsage = "N/A";
      let redisInfo = "N/A";
      try {
        const info = await redis.info("memory");
        const memoryMatch = info.match(/used_memory_human:([^\r\n]+)/);
        if (memoryMatch) {
          memoryUsage = memoryMatch[1].trim();
        }
        const serverInfo = await redis.info("server");
        const versionMatch = serverInfo.match(/redis_version:([^\r\n]+)/);
        if (versionMatch) {
          redisInfo = `Redis ${versionMatch[1].trim()}`;
        }
      } catch (e) {
      }
      return {
        totalKeys: keys.length,
        memoryUsage,
        redisInfo
      };
    },
    /**
     * Close the Redis connection
     */
    disconnect: async () => {
      await redis.disconnect();
    }
  };
  return store;
}
export {
  createRedisStore
};
