// toolbox/dist/storage/sqlite.js
import "../chunk-N3ETBM74.js";

// src/storage/sqlite.ts
import Database from "better-sqlite3";
import path from "path";
import os from "os";
function createSqliteStore(config = {}) {
  if (!Database) {
    throw new Error(
      "better-sqlite3 package not found. Please install it with: npm install better-sqlite3 @types/better-sqlite3"
    );
  }
  const {
    dbPath = path.join(os.tmpdir(), "mcp-store.db"),
    tableName = "mcp_store",
    enableWal = true,
    enableForeignKeys = false,
    timeout = 5e3,
    readonly = false
  } = config;
  const db = new Database(dbPath, {
    readonly,
    timeout
  });
  if (enableWal) {
    db.pragma("journal_mode = WAL");
  }
  if (enableForeignKeys) {
    db.pragma("foreign_keys = ON");
  }
  const createTableSql = `
    CREATE TABLE IF NOT EXISTS ${tableName} (
      key TEXT PRIMARY KEY,
      value TEXT NOT NULL,
      created_at DATETIME DEFAULT CURRENT_TIMESTAMP,
      updated_at DATETIME DEFAULT CURRENT_TIMESTAMP
    )
  `;
  const createIndexSql = `
    CREATE INDEX IF NOT EXISTS idx_${tableName}_created_at 
    ON ${tableName} (created_at)
  `;
  db.exec(createTableSql);
  db.exec(createIndexSql);
  const insertStmt = db.prepare(`
    INSERT OR REPLACE INTO ${tableName} (key, value, updated_at) 
    VALUES (?, ?, CURRENT_TIMESTAMP)
  `);
  const selectStmt = db.prepare(`
    SELECT value FROM ${tableName} WHERE key = ?
  `);
  const deleteStmt = db.prepare(`
    DELETE FROM ${tableName} WHERE key = ?
  `);
  const existsStmt = db.prepare(`
    SELECT 1 FROM ${tableName} WHERE key = ? LIMIT 1
  `);
  const keysStmt = db.prepare(`
    SELECT key FROM ${tableName} ORDER BY created_at
  `);
  const cleanupStmt = db.prepare(`
    DELETE FROM ${tableName} 
    WHERE created_at < datetime('now', '-' || ? || ' days')
  `);
  const store = {
    write: async (k, v) => {
      insertStmt.run(k, JSON.stringify(v));
    },
    read: async (k) => {
      const row = selectStmt.get(k);
      if (!row) {
        return null;
      }
      try {
        return JSON.parse(row.value);
      } catch (parseError) {
        throw new Error(
          `Failed to parse stored value for key "${k}": Invalid JSON`
        );
      }
    },
    /**
     * Delete a key from the store
     */
    delete: async (k) => {
      const result = deleteStmt.run(k);
      return result.changes > 0;
    },
    /**
     * Check if a key exists in the store
     */
    exists: async (k) => {
      const row = existsStmt.get(k);
      return row !== void 0;
    },
    /**
     * List all keys in the store (useful for debugging)
     */
    keys: async () => {
      const rows = keysStmt.all();
      return rows.map((row) => row.key);
    },
    /**
     * Clean up old entries (older than specified days)
     */
    cleanup: async (olderThanDays = 30) => {
      const result = cleanupStmt.run(olderThanDays);
      return result.changes;
    },
    /**
     * Get database statistics
     */
    stats: async () => {
      const countResult = db.prepare(`SELECT COUNT(*) as count FROM ${tableName}`).get();
      const sizeResult = db.prepare(
        "SELECT page_count * page_size as size FROM pragma_page_count(), pragma_page_size()"
      ).get();
      return {
        totalKeys: countResult.count,
        dbSize: `${(sizeResult.size / 1024).toFixed(2)} KB`,
        dbPath
      };
    },
    /**
     * Close the SQLite connection
     */
    disconnect: async () => {
      db.close();
    }
  };
  return store;
}
export {
  createSqliteStore
};
